File: clients/csharp/Model/Action.cs
namespace AiCup22.Model
{
    /// <summary>
    /// Action unit is currently performing
    /// </summary>
    public struct Action
    {
        /// <summary>
        /// Tick when the action will be finished
        /// </summary>
        public int FinishTick { get; set; }
        /// <summary>
        /// Type of the action
        /// </summary>
        public AiCup22.Model.ActionType ActionType { get; set; }
    
        public Action(int finishTick, AiCup22.Model.ActionType actionType)
        {
            this.FinishTick = finishTick;
            this.ActionType = actionType;
        }
    
        /// <summary> Read Action from reader </summary>
        public static Action ReadFrom(System.IO.BinaryReader reader)
        {
            var result = new Action();
            result.FinishTick = reader.ReadInt32();
            result.ActionType = ActionTypeHelper.ReadFrom(reader);
            return result;
        }
    
        /// <summary> Write Action to writer </summary>
        public void WriteTo(System.IO.BinaryWriter writer)
        {
            writer.Write(FinishTick);
            writer.Write((int) (ActionType));
        }
    
        /// <summary> Get string representation of Action </summary>
        public override string ToString() {
            string stringResult = "Action { ";
            stringResult += "FinishTick: ";
            stringResult += FinishTick.ToString();
            stringResult += ", ";
            stringResult += "ActionType: ";
            stringResult += ActionType.ToString();
            stringResult += " }";
            return stringResult;
        }
    }
}

File: clients/fsharp/Model/Action.fs
#nowarn "0058"

namespace AiCup22
namespace AiCup22.Model

open AiCup22

/// Action unit is currently performing
type Action = {
    /// Tick when the action will be finished
    FinishTick: int;
    /// Type of the action
    ActionType: Model.ActionType;
} with

    /// Write Action to writer
    member this.writeTo(writer: System.IO.BinaryWriter) =
        writer.Write this.FinishTick
        writer.Write (int this.ActionType)
        ()

    /// Read Action from reader
    static member readFrom(reader: System.IO.BinaryReader) = {
        FinishTick = reader.ReadInt32()
        ActionType = reader.ReadInt32() |> enum
    }